### PR TITLE
crypto modifications

### LINKED ISSUE #10317: Extend vigenère in crypto to accept repetitions in keys
In line 369 of function `encipher_vigenere`, which is part of `crypto.py` the method `uniq` from utilities.iterables is used to return an iterator of the key-phrase without possible spaces in the key. However, `uniq` will return an iterator with only unique characters of the key. Therefore, if one uses a key such as "cool", `uniq` returns "col" and the function will hence encrypt in a wrong way.

Example: `encipher_vigenere("wow", "cool")` returns "YCH". However, the right result would actually be "YCK".

Also, instead of using `capitalize` on each character of the key or the plaintext (see lines 370 or 373) we could also simplify the problem by using `key.upper` or `pt.upper`.

I would volunteer to fix this and if I am assuming right and `uniq` is used wrong also in other crypto functions such as `encipher_bifid5` or `bifid5_square` I could also fix these functions.

### PR DESCRIPTION
**Pull request by smichr**
modifications to the crypto module
- edited docstrings
- more consistent arg handling
- additional functionality
- a `replace` and `translate` method added to the `misc` library; `translate` uses `replace` and tries to be an efficient method to handle string replacements when they might have a mix of unicode and/or multicharacter patterns.

closes #10317 
closes #10328 by inclusion here

### CODE REVIEW

**Comment by smichr** [comment]
> Why does one have to import AZ here

no longer needed. In latter commits I clean up arg handling. I see that I have some more cleanup of doctests to do. Also, the `rst` file needs updating to clear the testing error.

**Comment by smichr** [comment]
Laguna has to be indented, apparently.

**Comment by smichr** [comment]
@ec-m , @wdjoyner do you have anny feedback about the changes?

**Comment by wdjoyner** [comment]
On Sun, Jan 24, 2016 at 9:20 PM, Christopher Smith <notifications@github.com
> wrote:

> @ec-m <https://github.com/ec-m> , @wdjoyner <https://github.com/wdjoyner>
> do you have anny feedback about the changes?
>

Can you remind me how to access your crypto.py module?

**Comment by ec-m** [comment]
@wdjoyner You should be able to access it here: https://github.com/smichr/sympy/blob/crypto/sympy/crypto/crypto.py

@smichr I think this is a nice makeover. I tested vigenère, hill, elgamal and diffie-hellman and they all worked as expected.

**Comment by wdjoyner** [comment]
On Mon, Jan 25, 2016 at 6:20 AM, Eva Charlotte Mayer <
notifications@github.com> wrote:

> @wdjoyner <https://github.com/wdjoyner> You should be able to access it
> here: https://github.com/smichr/sympy/blob/crypto/sympy/crypto/crypto.py
>

Nice. Thanks, Chris!

Minor nitpicks for the docstrings (which are possibly my fault):

line 60: I would replace "Return a string of the characters of ``symbols``
with those of" by "Return a string of the distinct characters of
``symbols`` with those of"

line 100: Replace "symbols: iterable of characters allowed in phrase; if
None," by "symbols: iterable of distinct characters allowed in phrase; if
None,"

line 148: Replace "Returns the cyclic shift of the list range(n) by
k." by "Returns
the cyclic shift to the left of the list range(n) by k."

line 184: Replace "``key``: an integer from (the secret key)" by "``key``:
an integer (the secret key)"

lines 186, 194: replace "m" by "msg"

line 226: Missing docstring. Perhaps you could use "See the method
encipher_shift."
?

Line 323: Code for encipher_substitution: Is the translate function the
same as the translate method (
https://docs.python.org/2/library/stdtypes.html#str.translate)? If not,
should this be referenced in the docstring?




@smichr <https://github.com/smichr> I think this is a nice makeover. I

**Comment by smichr** [comment]
> line 100: Replace "symbols: iterable of characters

The symbols in `symbols` need not be unique for the `check_and_join` step; they do, however, for the `padded_key` function since if there are duplicates at that step, a unique key can be returned but it would be a little less transparent than if the user decided on how to make the symbols unique.

So I didn't modify the text at line 100. I did modify it at line 60 (though doing so is redundant since an error is already raised if there are duplicates).

Thanks to both of you for taking time to look this over. I appreciated getting more familiar with this code.  I'll wait another 24 hrs to see if there are any more comments and then commit.

**Comment by smichr** [comment]
- [ ]  fix type rotine
- [ ] add note about translate which is patterned after the string method

File: sympy/utilities/misc.py [hunk 1]
```diff
@@ -217,5 +219,183 @@ def find_executable(executable, path=None):
...
+    return _replace(reps)(string)
+
+
+def translate(s, a, b=None, c=None):
+    """Return ``s`` where characters have been replaced or deleted.
+
+    SYNTAX
+    ======
+
+    translate(None, deletechars):
```
Comment (smichr) [inline sympy/utilities/misc.py]: add the `s` argument and reference in SYNTAX notes, e.g. `translate(s, None, deletechars)`

File: sympy/utilities/misc.py [hunk 2]
```diff
@@ -217,5 +219,183 @@ def find_executable(executable, path=None):
...
+    >>> translate(abc, {'a': 'x'}, 'c')
+    'xb'
+    >>> translate(abc, {'abc': 'x', 'a': 'y'})
+    'x'
+
+    >>> translate('abcd', 'ac', 'AC', 'd')
+    'AbC'
+
+    There is no guarantee that a unique answer will be
+    obtained if keys in a mapping overlap are the same
```
Comment (smichr) [inline sympy/utilities/misc.py]: ~~overlap~~

**Comment by smichr** [comment]
I just noticed that you can edit the source online -- that's a nice feature for quick touch-ups.

**Pull request merged**